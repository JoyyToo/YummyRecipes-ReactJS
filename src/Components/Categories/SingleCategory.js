import React, {Component} from 'react';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import {Card, CardText,} from 'material-ui/Card';
import FlatButton from 'material-ui/FlatButton';
import {Link} from 'react-router-dom';
import {notify} from 'react-notify-toast'
import axios from 'axios';
import * as constant from "../constant";

class SingleCategory extends Component {
    // initialize state
    constructor(props) {
        super(props);
        this.state = {
            categories: [],
            id: "",
            date_created: '',
            date_modified: '',
            desc: '',
            name: '',

        }
    }


    // mount token when page loads
    componentDidMount() {
        this.setState({
            token: window.sessionStorage.accessToken,
        });
        this.handlecategory();
        const token = window.localStorage.getItem('token');
        if (!token) {
            window.location.replace('/login')
        }
    }

    // handle get category request
    handlecategory = (event) => {
        const id = this.props.match.params['id'];

        const token = window.localStorage.getItem('token');
        // send GET request to API
        axios({
            url: `${constant.URL}/category/${id}`,
            method: 'get',
            headers: {
                Authorization: `Bearer ${token}`,
                'Content-Type': 'multipart/form-data',
            }

        }).then((response) => {
            let categories = response.data['category'];

            this.setState({
                date_created: categories['date_created'],
                date_modified: categories['date_modified'],
                desc: categories['desc'],
                name: categories['name'],
                id: categories['id'],
            })
        })

        .catch((error) => {
            notify.show(error.response, 'error', 4000);
        });
    };

    // handle delete category request
    handleDeletecategory = (event) => {

        let id = event.currentTarget.getAttribute('id');

        let th = this;

        const token = window.localStorage.getItem('token');
        // send DELETE request to API
        axios({
            url: `${constant.URL}/category/${id}`,
            method: 'delete',
            headers: {
                Authorization: `Bearer ${token}`,
                'Content-Type': 'multipart/form-data',
            }
        })

            .then((response) => {

                th.handlecategory();

                notify.show(response.data.message, 'success', 4000);
                this.setState({
                    id: this.state.id
                });
                this.props
                    .history
                    .push('/categories');
            })

            .catch((error) => {
                notify.show(error.response.data.message, 'error', 4000);
            });
    };


    // render a single category
    render() {
        let category = this.state;

        const styles = {
            card2: {
                position: 'relative',
                padding: 10,
                width: 450,
                float: 'left',
                margin: 5,

            },
            align: {
                align: 'left'

            }
        };
        return (
            <div>
                <MuiThemeProvider>

                    <div>
                        {(
                            <div style={styles.card2}>
                                <MuiThemeProvider is="nospace start">
                                    <Card >
                                        <h2 style={{marginLeft: 30, paddingTop: 10}}>{category['name']}</h2>
                                        <p style={{marginLeft: 30, marginBottom: 30}}>{category['desc']}</p>
                                        <CardText>
                                            {
                                                <div>

                                                    <p style={{marginLeft: 15}}><i>Created:</i> {this.state.date_created}</p>
                                                    <p style={{marginLeft: 15}}><i>Modified:</i> {this.state.date_modified}</p>

                                                </div>

                                            }
                                        </CardText>
                                        <Link to={"/updatecategory/" + category['id']}>
                                            <FlatButton label="Edit"
                                                        id={category['id']} desc={category['desc']}
                                                        primary={true}/></Link>
                                        <FlatButton label="Delete" style={{color: 'red'}}
                                                    id={category['id']}
                                                    onClick={this.handleDeletecategory.bind(this)}
                                        />

                                        <Link to={"/categories/" + category['id'] + "/recipes"}><FlatButton
                                                label="View Items"/></Link><br/>
                                    </Card>

                                </MuiThemeProvider>
                            </div>
                        )}
                    </div>

                </MuiThemeProvider>
            </div>

        )
    }

};

export default SingleCategory;

