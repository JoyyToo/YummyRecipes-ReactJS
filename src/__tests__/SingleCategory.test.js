import React from 'react';
import { shallow, mount} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import SingleCategory from '../Components/Categories/SingleCategory'

describe('SingleCategory component', () => {
    const params = {
        match: {
            params: {
                id:1
            }
        }
    }
    const wrapper = shallow(<SingleCategory match={{params}}/>);

    it('renders properly', () => {
        expect(shallowToJson(wrapper)).toMatchSnapshot();
      });
  
    it('renders 1 <SingleCategory/> component', () =>{
        const component = shallow(<SingleCategory match={{params}}/>);
        expect(component).toHaveLength(1);
    });
  
    it('it returns props correctly', () =>{
        const component = shallow(<SingleCategory name="app" match={{params}}/>);
        expect(component.instance().props.name).toBe('app')
    });

    it('contains MuiThemeProvider', () => {
        expect(wrapper.find('MuiThemeProvider')).toHaveLength(2);
    }); 

    it('contains divs', () => {
        expect(wrapper.find('div')).toHaveLength(4);
    });

    it('contains Card', () => {
        expect(wrapper.find('Card')).toHaveLength(1);
    });

    it('contains h2', () => {
        expect(wrapper.find('h2')).toHaveLength(1);
    });

    it('contains p', () => {
        expect(wrapper.find('p')).toHaveLength(3);
    });

    it('contains CardText', () => {
        expect(wrapper.find('CardText')).toHaveLength(1);
    });

    it('containsLink', () => {
        expect(wrapper.find('Link')).toHaveLength(2);
    });
    
})