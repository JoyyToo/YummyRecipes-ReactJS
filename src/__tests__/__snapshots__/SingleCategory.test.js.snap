// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleCategory component renders properly 1`] = `
<div>
  <MuiThemeProvider>
    <div>
      <div
        style={
          Object {
            "float": "left",
            "margin": 5,
            "padding": 10,
            "position": "relative",
            "width": 450,
          }
        }
      >
        <MuiThemeProvider
          is="nospace start"
        >
          <Card
            expandable={false}
            expanded={null}
            initiallyExpanded={false}
          >
            <h2
              style={
                Object {
                  "marginLeft": 30,
                  "paddingTop": 10,
                }
              }
            />
            <p
              style={
                Object {
                  "marginBottom": 30,
                  "marginLeft": 30,
                }
              }
            />
            <CardText>
              <div>
                <p
                  style={
                    Object {
                      "marginLeft": 15,
                    }
                  }
                >
                  <i>
                    Created: 
                  </i>
                </p>
              </div>
            </CardText>
            <Link
              replace={false}
              to="/updatecategory/"
            >
              <FlatButton
                desc=""
                disabled={false}
                fullWidth={false}
                id=""
                label="Edit"
                labelPosition="after"
                labelStyle={Object {}}
                onKeyboardFocus={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onTouchStart={[Function]}
                primary={true}
                secondary={false}
              />
            </Link>
            <FlatButton
              disabled={false}
              fullWidth={false}
              id=""
              label="Delete"
              labelPosition="after"
              labelStyle={Object {}}
              onClick={[Function]}
              onKeyboardFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onTouchStart={[Function]}
              primary={false}
              secondary={false}
              style={
                Object {
                  "color": "red",
                }
              }
            />
            <Link
              replace={false}
              to="/categories//recipes"
            >
              <FlatButton
                disabled={false}
                fullWidth={false}
                label="RECIPES"
                labelPosition="after"
                labelStyle={Object {}}
                onKeyboardFocus={[Function]}
                onMouseEnter={[Function]}
                onMouseLeave={[Function]}
                onTouchStart={[Function]}
                primary={false}
                secondary={false}
              />
            </Link>
            <br />
          </Card>
        </MuiThemeProvider>
      </div>
    </div>
  </MuiThemeProvider>
</div>
`;
